pkg/kubectl/resource/mapper.go

```diff
@@ -54,19 +56,21 @@ func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
 	if err != nil {
 		return nil, fmt.Errorf("unable to decode %q: %v", source, err)
 	}
-	var obj runtime.Object
-	var versioned runtime.Object
-	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
-		return nil, err
-	} else if isThirdParty {
-		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
-		versioned = obj
-		gvk = gvkOut
-	} else {
-		obj, versioned = versions.Last(), versions.First()
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
+
+	obj, versioned := versions.Last(), versions.First()
+	if !m.dynamic {
```

Comment:
That could be DynamicClientForMapping.  I don't know that `dynamic` is the
best word though going forward - you really want an unstructured client.
All third party objects are unstructured as well.   UnstructuredObject()
and UnstructuredClientForMapping() seem very reasonable to have on factory.

On Tue, Aug 9, 2016 at 2:44 PM, krousey <notifications@github.com> wrote:

> In pkg/kubectl/resource/mapper.go
> <https://github.com/kubernetes/kubernetes/pull/30250#discussion_r74120315>
> :
>
> > -	var obj runtime.Object
> > -	var versioned runtime.Object
> > -	if isThirdParty, gvkOut, err := thirdpartyresourcedata.IsThirdPartyObject(data, gvk); err != nil {
> > -		return nil, err
> > -	} else if isThirdParty {
> > -		obj, err = runtime.Decode(thirdpartyresourcedata.NewDecoder(nil, gvkOut.Kind), data)
> > -		versioned = obj
> > -		gvk = gvkOut
> > -	} else {
> > -		obj, versioned = versions.Last(), versions.First()
> > -	}
> > -	if err != nil {
> > -		return nil, fmt.Errorf("unable to decode %q: %v [%v]", source, err, gvk)
> > +
> > +	obj, versioned := versions.Last(), versions.First()
> > +	if !m.dynamic {
>
> The codecs provided by the factory come from ClientForMapping whose only
> input is the GVK. There is code in there to handle third party based on the
> GVK and use a third party codec instead.
>
> My needs are slightly different. If the caller wants to use dynamic
> decoding, that is the only codec they should use, regardless of the GVK.
> But I also need to leave the orginal codecs intact for commands that can't
> use the dynamic objects yet. (like describe, delete, ....). So the choice
> to use dynamic objects here isn't based on the mapping, but rather if the
> caller doesn't need the versioned types.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30250/files/8622d5ce64b4c381409b5fe86351e232b5e8e817#r74120315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-QDr2h1iyIEBd9Nyqjaw31CJLgMks5qeMqdgaJpZM4Jfj_0>
> .
>